File: frontend/src/api/issue.ts
/**
 * Issue API - 工作项 API 调用模块
 */
import api from './index'
import type {
  IssueCreate,
  IssueUpdate,
  IssueResponse,
  IssueSearchResult,
  IssueActivity,
  IssueStatistics,
  IssuePriority
} from '@/types/issue'

// ==================== Issue CRUD ====================

/**
 * 创建工作项
 */
export async function createIssue(
  projectId: number,
  workspaceId: number,
  data: IssueCreate
): Promise<IssueResponse> {
  const response = await api.post(
    `/issues/?project_id=${projectId}&workspace_id=${workspaceId}`,
    data
  )
  return response.data
}

/**
 * 列出项目的工作项（带分页信息）
 */
export interface IssueListResult {
  items: IssueResponse[]
  total: number
}

export async function listIssues(
  projectId: number,
  workspaceId: number,
  filters?: {
    state_id?: number
    priority?: IssuePriority
    assignee_id?: number
    parent_id?: number
    cycle_id?: number
    module_id?: number
    search?: string
    is_draft?: boolean
    issue_type_id?: number
    cf_field_id?: number
    cf_value?: string
    cf_and?: string
    limit?: number
    offset?: number
  }
): Promise<IssueListResult> {
  const params = new URLSearchParams()
  params.append('project_id', projectId.toString())
  params.append('workspace_id', workspaceId.toString())
  
  if (filters) {
    if (filters.state_id) params.append('state_id', filters.state_id.toString())
    if (filters.priority) params.append('priority', filters.priority)
    if (filters.assignee_id) params.append('assignee_id', filters.assignee_id.toString())
    if (filters.parent_id) params.append('parent_id', filters.parent_id.toString())
    if (filters.cycle_id) params.append('cycle_id', filters.cycle_id.toString())
    if (filters.module_id) params.append('module_id', filters.module_id.toString())
    if (filters.search) params.append('search', filters.search)
    if (filters.is_draft !== undefined) params.append('is_draft', filters.is_draft.toString())
    if (filters.issue_type_id) params.append('issue_type_id', filters.issue_type_id.toString())
    if (filters.cf_field_id) params.append('cf_field_id', filters.cf_field_id.toString())
    if (filters.cf_value) params.append('cf_value', filters.cf_value)
    if (filters.cf_and) params.append('cf_and', filters.cf_and)
    if (filters.limit) params.append('limit', filters.limit.toString())
    if (filters.offset) params.append('offset', filters.offset.toString())
  }
  
  const response = await api.get(`/issues/?${params.toString()}`)
  const total = parseInt(response.headers['x-total-count'] || '0', 10)
  return { items: response.data || [], total }
}

/**
 * 获取工作项详情
 */
export async function getIssue(issueId: number): Promise<IssueResponse> {
  const response = await api.get(`/issues/${issueId}`)
  return response.data
}

/**
 * 更新工作项
 */
export async function updateIssue(
  issueId: number,
  data: IssueUpdate
): Promise<IssueResponse> {
  const response = await api.put(`/issues/${issueId}`, data)
  return response.data
}

/**
 * 删除工作项
 */
export async function deleteIssue(issueId: number): Promise<void> {
  await api.delete(`/issues/${issueId}`)
}

/**
 * 归档工作项
 */
export async function archiveIssue(issueId: number): Promise<void> {
  await api.post(`/issues/${issueId}/archive`)
}

/**
 * 恢复工作项
 */
export async function restoreIssue(issueId: number): Promise<IssueResponse> {
  const response = await api.post(`/issues/${issueId}/restore`)
  return response.data
}

// ==================== Issue Activities ====================

/**
 * 获取工作项活动历史
 */
export async function getIssueActivities(
  issueId: number,
  limit?: number,
  offset?: number
): Promise<IssueActivity[]> {
  const params = new URLSearchParams()
  if (limit) params.append('limit', limit.toString())
  if (offset) params.append('offset', offset.toString())
  
  const response = await api.get(`/issues/${issueId}/activities?${params.toString()}`)
  return response.data
}

// ==================== Issue Statistics ====================

/**
 * 获取项目工作项统计
 */
export async function getIssueStatistics(projectId: number): Promise<IssueStatistics> {
  const response = await api.get(`/issues/statistics?project_id=${projectId}`)
  return response.data
}

// ==================== Issue Search ====================

/**
 * 搜索工作项
 */
export async function searchIssues(
  workspaceId: number,
  query: string,
  projectId?: number,
  limit?: number
): Promise<IssueSearchResult[]> {
  const params = new URLSearchParams()
  params.append('workspace_id', workspaceId.toString())
  params.append('query', query)
  
  if (projectId) params.append('project_id', projectId.toString())
  if (limit) params.append('limit', limit.toString())
  
  const response = await api.get(`/issues/search?${params.toString()}`)
  return response.data
}

// ==================== Bulk Operations ====================

/**
 * 批量更新工作项
 */
export async function bulkUpdateIssues(
  projectId: number,
  issueIds: number[],
  data: IssueUpdate
): Promise<IssueResponse[]> {
  const response = await api.post(
    `/issues/bulk/update?project_id=${projectId}`,
    { issue_ids: issueIds, ...data }
  )
  return response.data
}

/**
 * 批量删除工作项
 */
export async function bulkDeleteIssues(issueIds: number[]): Promise<void> {
  await api.post('/issues/bulk/delete', { issue_ids: issueIds })
}

export interface ImportError {
  row: number
  title: string
  message: string
}

export interface ImportResult {
  success_count: number
  fail_count: number
  errors: ImportError[]
  imported_ids: number[]
}

export interface ImportIssueItem {
  name: string
  description?: string
  priority?: string
  state_name?: string
  type_name?: string
  assignee_emails?: string[]
  label_names?: string[]
  start_date?: string
  target_date?: string
  parent_title?: string
}

/**
 * JSON 批量导入工作项
 */
export async function importIssuesJSON(
  projectId: number,
  workspaceId: number,
  items: ImportIssueItem[]
): Promise<ImportResult> {
  const response = await api.post(
    `/issues/import/json?project_id=${projectId}&workspace_id=${workspaceId}`,
    items
  )
  return response.data
}

/**
 * CSV 批量导入工作项
 */
export async function importIssuesCSV(
  projectId: number,
  workspaceId: number,
  file: File
): Promise<ImportResult> {
  const formData = new FormData()
  formData.append('file', file)
  const response = await api.post(
    `/issues/import/csv?project_id=${projectId}&workspace_id=${workspaceId}`,
    formData,
    {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    }
  )
  return response.data
}

// ==================== Assignee Management ====================

/**
 * 添加工作项负责人
 */
export async function addIssueAssignee(
  issueId: number,
  userId: number
): Promise<{ issue_id: number; user_id: number; action: string }> {
  const response = await api.post(`/issues/${issueId}/assignees?user_id=${userId}`)
  return response.data
}

/**
 * 移除工作项负责人
 */
export async function removeIssueAssignee(
  issueId: number,
  userId: number
): Promise<{ issue_id: number; user_id: number; action: string }> {
  const response = await api.delete(`/issues/${issueId}/assignees/${userId}`)
  return response.data
}

// ==================== Label Management ====================

/**
 * 添加工作项标签
 */
export async function addIssueLabel(
  issueId: number,
  labelId: number
): Promise<{ issue_id: number; label_id: number; action: string }> {
  const response = await api.post(`/issues/${issueId}/labels?label_id=${labelId}`)
  return response.data
}

/**
 * 移除工作项标签
 */
export async function removeIssueLabel(
  issueId: number,
  labelId: number
): Promise<{ issue_id: number; label_id: number; action: string }> {
  const response = await api.delete(`/issues/${issueId}/labels/${labelId}`)
  return response.data
}

// ==================== Cycle Management ====================

/**
 * 设置工作项周期
 */
export async function setIssueCycle(
  issueId: number,
  cycleId: number
): Promise<{ issue_id: number; cycle_id: number; action: string }> {
  const response = await api.post(`/issues/${issueId}/cycle?cycle_id=${cycleId}`)
  return response.data
}

/**
 * 移除工作项周期
 */
export async function removeIssueCycle(
  issueId: number
): Promise<{ issue_id: number; cycle_id: null; action: string }> {
  const response = await api.delete(`/issues/${issueId}/cycle`)
  return response.data
}

// ==================== Pages ====================

export async function listIssuePages(issueId: number): Promise<any[]> {
  const response = await api.get(`/issues/${issueId}/pages`)
  return response.data
}

export async function addIssuePage(issueId: number, pageId: number): Promise<any> {
  const response = await api.post(`/issues/${issueId}/pages?page_id=${pageId}`)
  return response.data
}

export async function removeIssuePage(issueId: number, pageId: number): Promise<any> {
  const response = await api.delete(`/issues/${issueId}/pages?page_id=${pageId}`)
  return response.data
}

// ==================== Export all ====================

export const issueApi = {
  // CRUD
  createIssue,
  listIssues,
  getIssue,
  updateIssue,
  deleteIssue,
  archiveIssue,
  restoreIssue,
  
  // Activities
  getIssueActivities,
  
  // Statistics
  getIssueStatistics,
  
  // Search
  searchIssues,
  
  // Bulk
  bulkUpdateIssues,
  bulkDeleteIssues,
  importIssuesJSON,
  importIssuesCSV,
  
  // Assignee
  addIssueAssignee,
  removeIssueAssignee,
  
  // Label
  addIssueLabel,
  removeIssueLabel,
  
  // Cycle
  setIssueCycle,
  removeIssueCycle,
  
  // Pages
  listIssuePages,
  addIssuePage,
  removeIssuePage
}

export default issueApi

File: backend/app/api/v1/endpoints/issue.py
"""
Issue API Endpoints - 工作项管理接口
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional, List

from app.db.session import get_db
from app.api.deps import get_current_user
from app.models.user import User
from app.models.project import Project
from app.schemas.issue import (
    IssueCreate,
    IssueUpdate,
    IssueResponse,
    IssueLite,
    IssueSearchResult,
    IssuePriority
)
from app.services import issue as issue_service
from app.core.exceptions import NotFoundException, ValidationException

router = APIRouter()


# ==================== Issue CRUD ====================

@router.post("/", response_model=IssueResponse, status_code=201)
async def create_issue(
    project_id: int,
    workspace_id: int,
    issue_data: IssueCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    创建工作项
    
    在指定项目中创建一个新的工作项（Issue、Task、Bug等）。
    可以设置名称、描述、优先级、状态、负责人、标签等属性。
    """
    # 验证项目访问权限
    project = await db.get(Project, project_id)
    if not project or project.is_deleted:
        raise HTTPException(status_code=404, detail="Project not found")
    
    issue = await issue_service.create_issue(
        db=db,
        issue_data=issue_data,
        project_id=project_id,
        workspace_id=workspace_id,
        user_id=current_user.id
    )
    
    return issue_service.build_issue_response(issue)


@router.get("/", response_model=List[IssueResponse])
async def list_issues(
    project_id: int,
    workspace_id: int,
    state_id: Optional[int] = None,
    priority: Optional[IssuePriority] = None,
    assignee_id: Optional[int] = None,
    parent_id: Optional[int] = None,
    cycle_id: Optional[int] = None,
    module_id: Optional[int] = None,
    search: Optional[str] = None,
    is_draft: Optional[bool] = None,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    列出项目的工作项
    
    支持多种筛选条件：
    - state_id: 按状态筛选
    - priority: 按优先级筛选
    - assignee_id: 按负责人筛选
    - parent_id: 按父工作项筛选
    - cycle_id: 按周期筛选
    - module_id: 按模块筛选
    - search: 搜索关键词
    - is_draft: 是否为草稿
    """
    issues = await issue_service.list_project_issues(
        db=db,
        project_id=project_id,
        state_id=state_id,
        priority=priority,
        assignee_id=assignee_id,
        parent_id=parent_id,
        cycle_id=cycle_id,
        module_id=module_id,
        search=search,
        is_draft=is_draft,
        limit=limit,
        offset=offset
    )
    
    return [issue_service.build_issue_response(issue) for issue in issues]


@router.get("/{issue_id}", response_model=IssueResponse)
async def get_issue(
    issue_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取工作项详情
    
    返回工作项的完整信息，包括关联的项目、状态、自定义字段值等。
    """
    issue = await issue_service.get_issue_by_id(db, issue_id)
    return issue_service.build_issue_response(issue)


@router.put("/{issue_id}", response_model=IssueResponse)
async def update_issue(
    issue_id: int,
    update_data: IssueUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    更新工作项
    
    可以更新工作项的名称、描述、优先级、状态、负责人、标签等属性。
    每次更新都会记录活动历史。
    """
    issue = await issue_service.update_issue(
        db=db,
        issue_id=issue_id,
        update_data=update_data,
        user_id=current_user.id
    )
    return issue_service.build_issue_response(issue)


@router.delete("/{issue_id}", status_code=204)
async def delete_issue(
    issue_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    删除工作项（软删除）
    
    工作项不会真正删除，而是标记为已删除状态。
    可以通过恢复接口恢复。
    """
    await issue_service.delete_issue(db, issue_id)
    return None


@router.post("/{issue_id}/archive", status_code=204)
async def archive_issue(
    issue_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    归档工作项
    
    归档的工作项会从常规列表中隐藏，但仍可通过归档列表查看。
    """
    await issue_service.archive_issue(db, issue_id)
    return None


@router.post("/{issue_id}/restore", response_model=IssueResponse)
async def restore_issue(
    issue_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    恢复工作项
    
    恢复已删除或归档的工作项。
    """
    issue = await issue_service.restore_issue(db, issue_id)
    return issue_service.build_issue_response(issue)


# ==================== Issue Activities ====================

@router.get("/{issue_id}/activities", response_model=List[dict])
async def get_issue_activities(
    issue_id: int,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取工作项活动历史
    
    返回工作项的所有变更记录，包括创建、更新、状态变更等。
    """
    activities = await issue_service.get_issue_activities(
        db=db,
        issue_id=issue_id,
        limit=limit,
        offset=offset
    )
    
    return [
        {
            "id": activity.id,
            "issue_id": activity.issue_id,
            "verb": activity.verb,
            "field": activity.field,
            "old_value": activity.old_value,
            "new_value": activity.new_value,
            "comment": activity.comment,
            "actor_id": activity.actor_id,
            "created_at": activity.created_at
        }
        for activity in activities
    ]


# ==================== Issue Statistics ====================

@router.get("/statistics", response_model=dict)
async def get_issue_statistics(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取项目工作项统计
    
    返回工作项的总数、按状态分组统计、按优先级分组统计等。
    """
    stats = await issue_service.get_issue_statistics(db, project_id)
    return stats


# ==================== Issue Search ====================

@router.get("/search", response_model=List[IssueSearchResult])
async def search_issues(
    workspace_id: int,
    query: str = Query(..., min_length=1),
    project_id: Optional[int] = None,
    limit: int = Query(10, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    搜索工作项
    
    在工作空间或项目范围内搜索工作项，支持名称和描述搜索。
    """
    # TODO: 实现跨项目搜索
    if project_id:
        issues = await issue_service.list_project_issues(
            db=db,
            project_id=project_id,
            search=query,
            limit=limit
        )
        
        return [
            IssueSearchResult(
                id=issue.id,
                name=issue.name,
                sequence_id=issue.sequence_id,
                project_identifier=issue.project.identifier if issue.project else "",
                project_id=issue.project_id,
                workspace_slug=workspace_id  # TODO: 获取 workspace slug
            )
            for issue in issues
        ]
    
    return []


# ==================== Bulk Operations ====================

@router.post("/bulk/update", response_model=List[IssueResponse])
async def bulk_update_issues(
    project_id: int,
    issue_ids: List[int],
    update_data: IssueUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    批量更新工作项
    
    同时更新多个工作项的属性（如状态、优先级、负责人等）。
    """
    results = []
    for issue_id in issue_ids:
        issue = await issue_service.update_issue(
            db=db,
            issue_id=issue_id,
            update_data=update_data,
            user_id=current_user.id
        )
        results.append(issue_service.build_issue_response(issue))
    
    return results


@router.post("/bulk/delete", status_code=204)
async def bulk_delete_issues(
    issue_ids: List[int],
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    批量删除工作项
    """
    for issue_id in issue_ids:
        await issue_service.delete_issue(db, issue_id)
    
    return None


# ==================== Assignee Management ====================

@router.post("/{issue_id}/assignees", response_model=dict)
async def add_issue_assignee(
    issue_id: int,
    user_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    添加工作项负责人
    
    将指定用户添加为工作项的负责人。
    """
    result = await issue_service.add_assignee(
        db=db,
        issue_id=issue_id,
        user_id=user_id,
        actor_id=current_user.id
    )
    return result


@router.delete("/{issue_id}/assignees/{user_id}", response_model=dict)
async def remove_issue_assignee(
    issue_id: int,
    user_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    移除工作项负责人
    
    从工作项的负责人列表中移除指定用户。
    """
    result = await issue_service.remove_assignee(
        db=db,
        issue_id=issue_id,
        user_id=user_id,
        actor_id=current_user.id
    )
    return result


# ==================== Label Management ====================

@router.post("/{issue_id}/labels", response_model=dict)
async def add_issue_label(
    issue_id: int,
    label_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    添加工作项标签
    
    将指定标签添加到工作项。
    """
    result = await issue_service.add_label(
        db=db,
        issue_id=issue_id,
        label_id=label_id,
        actor_id=current_user.id
    )
    return result


@router.delete("/{issue_id}/labels/{label_id}", response_model=dict)
async def remove_issue_label(
    issue_id: int,
    label_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    移除工作项标签
    
    从工作项的标签列表中移除指定标签。
    """
    result = await issue_service.remove_label(
        db=db,
        issue_id=issue_id,
        label_id=label_id,
        actor_id=current_user.id
    )
    return result


# ==================== Cycle Management ====================

@router.post("/{issue_id}/cycle", response_model=dict)
async def set_issue_cycle(
    issue_id: int,
    cycle_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    设置工作项周期
    
    将工作项添加到指定的迭代周期。
    """
    result = await issue_service.set_issue_cycle(
        db=db,
        issue_id=issue_id,
        cycle_id=cycle_id,
        actor_id=current_user.id
    )
    return result


@router.delete("/{issue_id}/cycle", response_model=dict)
async def remove_issue_cycle(
    issue_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    移除工作项周期
    
    将工作项从当前周期中移除。
    """
    result = await issue_service.remove_issue_cycle(
        db=db,
        issue_id=issue_id,
        actor_id=current_user.id
    )
    return result